From 32036e2a4ee2f813734ae63c83dd601c5c409cee Mon Sep 17 00:00:00 2001
From: Asahi Lina <lina@asahilina.net>
Date: Tue, 1 Jul 2025 16:48:12 +0900
Subject: [PATCH 1/5] frontend: Consider settings changed if an output setting
 was reset

If the previous selection is no longer available, the UI would default
to the first index, but not allow the user to save the change (which
then becomes impossible to save if the setting only has one possible
value, since you can never trigger the "changed" status).

If the settings load process clamped any setting to the default, enable
the apply button and consider the setting changed.
---
 frontend/settings/OBSBasicSettings.cpp | 28 +++++++++++++++++++++++---
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/frontend/settings/OBSBasicSettings.cpp b/frontend/settings/OBSBasicSettings.cpp
index 6e47f93ac924..98998f3605cf 100644
--- a/frontend/settings/OBSBasicSettings.cpp
+++ b/frontend/settings/OBSBasicSettings.cpp
@@ -1753,19 +1753,30 @@ void OBSBasicSettings::LoadSimpleOutputSettings()
 	ui->simpleOutAdvanced->setChecked(advanced);
 	ui->simpleOutCustom->setText(custom);
 
+	bool outputSettingReset = false;
+
 	idx = ui->simpleOutRecQuality->findData(QString(recQual));
-	if (idx == -1)
+	if (idx == -1) {
+		ui->simpleOutRecQuality->setProperty("changed", QVariant(true));
 		idx = 0;
+		outputSettingReset = true;
+	}
 	ui->simpleOutRecQuality->setCurrentIndex(idx);
 
 	idx = ui->simpleOutStrEncoder->findData(QString(streamEnc));
-	if (idx == -1)
+	if (idx == -1) {
+		ui->simpleOutStrEncoder->setProperty("changed", QVariant(true));
 		idx = 0;
+		outputSettingReset = true;
+	}
 	ui->simpleOutStrEncoder->setCurrentIndex(idx);
 
 	idx = ui->simpleOutStrAEncoder->findData(QString(streamAudioEnc));
-	if (idx == -1)
+	if (idx == -1) {
+		ui->simpleOutStrAEncoder->setProperty("changed", QVariant(true));
 		idx = 0;
+		outputSettingReset = true;
+	}
 	ui->simpleOutStrAEncoder->setCurrentIndex(idx);
 
 	idx = ui->simpleOutRecEncoder->findData(QString(recEnc));
@@ -1781,6 +1792,10 @@ void OBSBasicSettings::LoadSimpleOutputSettings()
 	ui->simpleRBMegsMax->setValue(rbSize);
 
 	SimpleStreamingEncoderChanged();
+	if (outputSettingReset) {
+		outputsChanged = true;
+		EnableApplyButton(true);
+	}
 }
 
 static inline QString makeFormatToolTip()
@@ -1921,6 +1936,10 @@ void OBSBasicSettings::LoadAdvOutputStreamingEncoderProperties()
 
 			ui->advOutEncoder->insertItem(0, encName, QT_UTF8(type));
 			SetComboByValue(ui->advOutEncoder, type);
+		} else {
+			ui->advOutEncoder->setProperty("changed", QVariant(true));
+			outputsChanged = true;
+			EnableApplyButton(true);
 		}
 	}
 
@@ -2027,6 +2046,9 @@ void OBSBasicSettings::LoadAdvOutputRecordingEncoderProperties()
 			SetComboByValue(ui->advOutRecEncoder, type);
 		} else {
 			ui->advOutRecEncoder->setCurrentIndex(-1);
+			ui->advOutRecEncoder->setProperty("changed", QVariant(true));
+			outputsChanged = true;
+			EnableApplyButton(true);
 		}
 	}
 }
-- 
2.51.0

